### PR TITLE
Expanded documentation for tracing fields

### PR DESCRIPTION
**Pull request by sgryphon**
Additional usage documentation for tracing fields, with a longer explanation of trace.id, span.id, and how they are related, and the correlation purpose of each.

Documentation would not reproduce the diagrams inline, so added to the optional docs/usages/<field>.asciidoc file.

Addresses issue #998  

Note: I have added the generated files to the PR, as the build 'make check' wants them checked in.

[Doc page preview](https://ecs_1128.docs-preview.app.elstc.co/guide/en/ecs/master/ecs-tracing-usage.html)

### CODE REVIEW

**Comment by sgryphon** [comment]
@webmat I see you added the label 1.9.0, which I am guessing means version 1.9.   This is only additional documentation, not an actual schema change, so from SemVer point of view it is can be a patch version 1.7, as it is still fully compatible with 1.7

**Comment by webmat** [comment]
@sgryphon Yeah we did that tagging as part of regular planning and figuring what we have time for when. We hadn't noticed that you were actually contributing the documentation page itself.

This is great, thank you for submitting that :-)

We will review shortly, and we may indeed backport this all the way back to 1.7, so that the documentation page comes to life before the next release.

**Comment by webmat** [comment]
@elasticmachine, run elasticsearch-ci/docs

**Comment by webmat** [comment]
The doc page can now be previewed at https://ecs_1128.docs-preview.app.elstc.co/guide/en/ecs/master/ecs-tracing-usage.html :-)

**Comment by webmat** [comment]
@axw Do you feel like this doc page captures the discussion in #998?

**Review by axw — COMMENTED** [review]
Thanks @sgryphon! This looks pretty good overall.

I'm happy to have this usage page. There are some discrepancies with Elastic APM which, along with ECS loggers, is (AFAIK) the only thing using these fields at the moment. I think it's important that we make some adjustments to avoid confusion.

@felixbarny @bmorelli25 any thoughts, concerns?

File: docs/usage/tracing.asciidoc [hunk 1]
```diff
@@ -0,0 +1,62 @@
+[[ecs-tracing-usage]]
+==== Tracing Usage
+
+A distributed trace is a set of messages, triggered as a result of a single logical operation, consolidated across
```
[the remark this answers is not in the document]
Comment (sgryphon) [reply docs/usage/tracing.asciidoc]: Sounds great. Changed.

File: schemas/tracing.yml [hunk 2]
```diff
@@ -38,4 +41,7 @@
       description: >
         Unique identifier of the span within the scope of its trace.
 
-        A span represents an operation within a transaction, such as a request to another service, or a database query.
+        A span represents an operation within a trace, such as a request to another service, or a database query.
```
[the remark this answers is not in the document]
Comment (sgryphon) [reply schemas/tracing.yml]: Fixed

File: docs/usage/tracing.asciidoc [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+Temporal relationships between Spans in a single Trace
+
+––|–––––––|–––––––|–––––––|–––––––|–––––––|–––––––|–––––––|–> time
+
+[Span A···················································]
+[Span B··············································]
+    [Span D··········································]
+    [Span C········································]
+        [Span E·······]        [Span F··]
+------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (sgryphon) [reply docs/usage/tracing.asciidoc]: Updated the text to better align with APM and added the link to APM transactions. Modified the diagrams to indicate transaction spans (and service boundaries).

**Comment by sgryphon** [comment]
Pull request has been updated based on feedback. Please review and let me know if there is anything else needed before it can be merged. Thanks.

**Review by axw — DISMISSED** [review]
Thanks for the updates @sgryphon. LGTM!

**Comment by bmorelli25** [comment]
@elasticmachine, run elasticsearch-ci/docs

File: docs/usage/tracing.asciidoc [hunk 4]
```diff
@@ -0,0 +1,72 @@
+[[ecs-tracing-usage]]
+==== Tracing Usage
+
+A trace describes the individual operations, and their causality, that ensue as a result of a single logical
```
Comment (bmorelli25) [inline docs/usage/tracing.asciidoc:4]:
```suggestion
A trace describes the individual operations and their causality that ensue from a single logical 
```

File: docs/usage/tracing.asciidoc [hunk 5]
```diff
@@ -0,0 +1,72 @@
+[[ecs-tracing-usage]]
+==== Tracing Usage
+
+A trace describes the individual operations, and their causality, that ensue as a result of a single logical 
+operation. A distributed trace comprises operations across multiple distributed components, crossing process, 
+network, and security boundaries.
+
+For example, a distributed trace may be initiated when someone presses a button to start an action on a website - in this example,
```
Comment (bmorelli25) [inline docs/usage/tracing.asciidoc:8]: It's subtle, but this should be an em dash

```suggestion
For example, a distributed trace may be initiated when someone presses a button to start an action on a website – in this example, 
```

File: docs/usage/tracing.asciidoc [hunk 6]
```diff
@@ -0,0 +1,72 @@
+[[ecs-tracing-usage]]
+==== Tracing Usage
+
+A trace describes the individual operations, and their causality, that ensue as a result of a single logical 
+operation. A distributed trace comprises operations across multiple distributed components, crossing process, 
+network, and security boundaries.
+
+For example, a distributed trace may be initiated when someone presses a button to start an action on a website - in this example, 
+the trace will represent calls made between the downstream services that handled the chain of requests initiated by this
```
Comment (bmorelli25) [inline docs/usage/tracing.asciidoc:9]:
```suggestion
the trace represents calls made between the downstream services that handled the chain of requests initiated by this 
```

File: docs/usage/tracing.asciidoc [hunk 7]
```diff
@@ -0,0 +1,72 @@
+[[ecs-tracing-usage]]
+==== Tracing Usage
+
+A trace describes the individual operations, and their causality, that ensue as a result of a single logical 
+operation. A distributed trace comprises operations across multiple distributed components, crossing process, 
+network, and security boundaries.
+
+For example, a distributed trace may be initiated when someone presses a button to start an action on a website - in this example, 
+the trace will represent calls made between the downstream services that handled the chain of requests initiated by this 
+button being pressed.
```
Comment (bmorelli25) [inline docs/usage/tracing.asciidoc:10]:
```suggestion
button press.
```

File: docs/usage/tracing.asciidoc [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+A trace is made up of transactions and spans, where each transaction/span is one operation, activity, or part of the overall
+trace. A transaction is a special kind of span that is the highest level of work measured within a service. In Elastic APM, 
+transactions have additional attributes associated with them
```
Comment (bmorelli25) [inline docs/usage/tracing.asciidoc:14]:
```suggestion
transactions also have additional attributes associated with them.
```

File: docs/usage/tracing.asciidoc [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+network, and security boundaries.
+
+For example, a distributed trace may be initiated when someone presses a button to start an action on a website - in this example, 
+the trace will represent calls made between the downstream services that handled the chain of requests initiated by this 
+button being pressed.
+
+A trace is made up of transactions and spans, where each transaction/span is one operation, activity, or part of the overall
+trace. A transaction is a special kind of span that is the highest level of work measured within a service. In Elastic APM, 
+transactions have additional attributes associated with them 
+(for more detail see https://www.elastic.co/guide/en/apm/get-started/current/transactions.html).
```
Comment (bmorelli25) [inline docs/usage/tracing.asciidoc:15]: We use attributes for internal links. This ensures links are always correct -- even if pages move.

```suggestion
Learn more about transactions in {apm-overview-ref-v}/transactions.html[APM Transactions].
```

File: docs/usage/tracing.asciidoc [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+
+Usually a transaction corresponds to the top level request to a component. Sometimes a component may break up processing into
```
Comment (bmorelli25) [inline docs/usage/tracing.asciidoc:17]:
```suggestion
Usually, a transaction corresponds to the top-level request to a component. Sometimes a component may break up processing into 
```

File: docs/usage/tracing.asciidoc [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+A trace is made up of transactions and spans, where each transaction/span is one operation, activity, or part of the overall
+trace. A transaction is a special kind of span that is the highest level of work measured within a service. In Elastic APM, 
+transactions have additional attributes associated with them 
+(for more detail see https://www.elastic.co/guide/en/apm/get-started/current/transactions.html).
+
+Usually a transaction corresponds to the top level request to a component. Sometimes a component may break up processing into 
+multiple spans (and transactions) if needed. 
+
+The spans (and transactions) have a parent/child relationship with other spans and form a tree (or directed acylic grapph),
```
Comment (bmorelli25) [inline docs/usage/tracing.asciidoc:20]: Is this an either-or? I think it'd be more clear if we said "also known as" instead.

```suggestion
The spans (and transactions) have a parent/child relationship with other spans and form a tree (also known as a directed acyclic graph), 
```

File: docs/usage/tracing.asciidoc [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+A trace is made up of transactions and spans, where each transaction/span is one operation, activity, or part of the overall
+trace. A transaction is a special kind of span that is the highest level of work measured within a service. In Elastic APM, 
+transactions have additional attributes associated with them 
+(for more detail see https://www.elastic.co/guide/en/apm/get-started/current/transactions.html).
+
+Usually a transaction corresponds to the top level request to a component. Sometimes a component may break up processing into 
+multiple spans (and transactions) if needed. 
+
+The spans (and transactions) have a parent/child relationship with other spans and form a tree (or directed acylic grapph), 
+with each span (or transaction), except the root, having a parent.
```
Comment (bmorelli25) [inline docs/usage/tracing.asciidoc:21]:
```suggestion
with each span or transaction, except the root, having a parent.
```

File: docs/usage/tracing.asciidoc [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+The `trace.id` allows the distributed trace messages to be correlated across an entire logical operation.
+
+The `span.id` allows messages from a single section of processing within the overall trace, e.g. a request to a component, to be correlated together. The `transaction.id` is the equivalent for transaction spans.
+
+If following the https://www.w3.org/TR/trace-context/[W3C Trace Context Recommendation] then
```
Comment (bmorelli25) [inline docs/usage/tracing.asciidoc:27]:
```suggestion
If you're following the https://www.w3.org/TR/trace-context/[W3C Trace Context Recommendation],
```

File: docs/usage/tracing.asciidoc [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+the values should correspond to those specified in the recommendation and passed in the `traceparent` header, 
+with `trace.id` being 32 lowercase hex characters, and `span.id` being 16 lowercase hex characters.
+
+For correlation wtih Elastic Application Performance Monitoring (APM) see 
+the https://www.elastic.co/guide/en/apm/get-started/current/distributed-tracing.html[APM Distributed Tracing] documentation.
```
Comment (bmorelli25) [inline docs/usage/tracing.asciidoc:32]: We've already used APM on this page, so I think only using the initialism is acceptable.

```suggestion
For correlation with Elastic APM, see
{apm-overview-ref-v}/distributed-tracing.html[APM Distributed Tracing].
```

File: docs/usage/tracing.asciidoc [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+
+Service 1:
+[Transaction Span A·······································]
+[Span B··············································]
+    [Span D··········································]
+
+Service 2:
+    [Transaction Span C····························]
+        [Span E·······]        [Span F··]
+------------------------------------------------------------------
```
Comment (bmorelli25) [inline docs/usage/tracing.asciidoc:69]: I'm not a fan of how these render. Let me mock up some graphics to replace these. Give me an hour.

**Review by bmorelli25 — CHANGES_REQUESTED** [review]
Thanks again for putting this together! I have a few grammar/spelling recommendations. Everything is open for discussion, so feel free to let me know if you disagree with anything!

I'll mock up some graphics for this right now.

**Comment by bmorelli25** [comment]
Here's what I'm thinking for the causal and temporal graphics. Open to feedback.

causal-span-trace

temporal-span-trace

**Comment by ebeahan** [comment]
Thanks, @sgryphon, for taking on authoring this content!  The functionality to add the usage docs was added only recently, so it's great to see folks contributing already. 😄 

Also, thanks to @axw and @bmorelli25 for the reviews and feedback. [These graphics](https://github.com/elastic/ecs/pull/1128#issuecomment-767196715) look amazing! ❤️  💯 

From an ECS-perspective, I do have reservations that this content is too focused on terminology and concepts and not focused enough on how to use/populate the ECS fields themselves. I prefer directing users to the APM docs' relevant sections as the more definitive reference for definitions and concepts. The ECS usage docs can then focus on applying the concepts with real-world examples using the `tracing.*` fields.

Those types of extended examples will help tie together concepts better for ECS users when mapping their own data sources. Some of the original questions from https://github.com/elastic/ecs/issues/998 were "How do these [fields] work together?", and I want to ensure that remains in focus here.

**Comment by sgryphon** [comment]
> Here's what I'm thinking for the causal and temporal graphics. Open to feedback.
 
My question about the graphics are how are they generated?  Using ASCII is viewable even in text output.

Mostly for diagrams I use something like PlantUML, to specify the diagram in text (under version control) and then generate the image.

But if you prefer straight images that is okay. I will create some SVG of the able, which are then more easily trackable in version control for changes.

**Comment by sgryphon** [comment]
> Those types of extended examples will help tie together concepts better for ECS users when mapping their own data sources. Some of the original questions from #998 were "How do these [fields] work together?", and I want to ensure that remains in focus here.

Okay, I can expand on "how to populate the ECS fields".

I'll see what I can come up with for a general solution (e.g. users mapping their own sources), but implementations might end up being technology specific, e.g. how you would populate them in using the built in TraceContext support in .NET would be very different than if using the Elastic APM agent in Python, etc.

**Comment by bmorelli25** [comment]
Closing this and moving documentation to the APM Guide. Tracking in https://github.com/elastic/observability-docs/issues/817.

**Pull request closed without merging**